MRG: Use _Brain in plot_source_estimates

GuillaumeFavelier opened the pull request:
This PR continues the work of #6460 and #6515. The goal is to make `stc.plot()` use the `_Brain` class which relies on the abstract 3d backend layer `_Renderer`. Technically it provides a `PyVista` alternative to the `Mayavi`-based `PySurfer` or any backend that respects the `_Renderer` contract in general. 

In practice, the work has started with the `plot_source_estimates()` function.

***

**ToDo**

*- Ensure consistency of results among all the backends*
*- Update the `BaseRenderer` API*
*- Ensure proper testing of the `_Brain` features*

- [x] Add support for a colorbar and text
- [x] Fix time value in `text_label`
- [x] Import `mesh_edges` and `smoothing_matrix` from `PySurfer`
- [x] Update `stc` tests to use `renderer` fixture 
- [x] Add `add_label()` and `show_view()` functions
- [x] Add a `screenshot()` function
- [x] Add a `add_foci()` function
- [x] Add a `add_text()` function
- [x] Update branch with #6819
- [x] Add a `save_image()` function.

**ToDo in a follow-up PR**

- Port `TimeViewer` from `PySurfer`
- Add `remove_data()`, `remove_labels()`  and `remove_foci()` functions

**API**

`__init__()`:
- [x] Add support for `hemi=split`
- [ ] Add support for `cortex` and `geo_kwargs` *(on user need/request)*
- [x] Add support for `figure` ~~(naming by `int` id is missing)~~

`add_data()`:
- [x] Add support for `transparent`
- [ ] Add support for `verbose` *(on user need/request)*

`add_label()`:
- [ ] Add support for `label_name` *(useful for `remove_label`)*

`add_foci()`
- [ ] Add support for `map_surface` *(on user need/request)*

**Bugs**

- [x] The camera is not set up correctly (see this [comment](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-527885549))
- [x] The colorbar range is different than the reference (see this [comment](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-527912117), fixed by [370](https://github.com/pyvista/pyvista/pull/370))
- [x] Rework of colorbar colormap is [needed](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-527948272)
- [x] The colorbar should use [256 colors](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-528450738) (merged [upstream](https://github.com/pyvista/pyvista/pull/370))
- [x] The `text_label` position is different than the reference (see [here](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-528448452))
- [x] The font size of the colorbar ticks is smaller than the reference (see [here](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-528448452))
- [x] The label is displayed in the wrong layer (see [here](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-530853787))
- [x] Fix strong dependency to `nibabel` (Travis won't be happy until I fix this)
- [x] Fix bug with colormap (see [here](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-532611973))
- [x] Fix bug with `screenshot()` during doc building (merged [upstream](https://github.com/pyvista/pyvista/pull/382))
- [ ] Fix screenshot with `mode='rgba'` for the pyvista backend (`Plotter` class without `ren_win` attribute) *(on user need/request)*
- [x] Fix 'double colorbar' issue with `hemi=both`
- [x] Fix difference in cylinder radius for `quiver3d` (reported in this [comment](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-536042199)) *(magic number of 0.15 suggested in this [comment](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-537921373) did the trick)*
- [x] Fix [HiDPI issue](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-538001249) (merged [upstream](https://github.com/pyvista/pyvista/pull/390))

**Feature request**

Set the colorbar ticks (suggested in this [comment](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-527920541))
Improve overall lighting (suggested in this [comment](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-532625434))
Add linked views in case of different subjects/conditions (coming from this [comment](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-535002561))

GuillaumeFavelier commented:
Using `examples/visualization/plot_xhemi.py`:

Mayavi | PyVista
---------|-----------
![image](https://user-images.githubusercontent.com/18143289/64255971-bd1ad200-cf22-11e9-903b-c9593cf26797.png) | ![image](https://user-images.githubusercontent.com/18143289/64256047-e89dbc80-cf22-11e9-99f5-d1a936000455.png)


The colorbar is still missing and the camera point of view is not set up correctly:
![image](https://user-images.githubusercontent.com/18143289/64256039-e0458180-cf22-11e9-90bf-a41175a43b57.png)

agramfort commented:
nice progress ! you've never been so close

>

GuillaumeFavelier commented:
Using only the scalar values with the given colormap we obtain something completely different:

![image](https://user-images.githubusercontent.com/18143289/64260832-d83e0f80-cf2b-11e9-842a-bdde2516cc96.png)

**Note**

It seems that using a lookup table is the way to go for now. The colorbar has the decent colormap but the range is not correct.

GuillaumeFavelier commented:
This is the result by forcing the RGBA values with the lookup table and setting the scalar values (used for the colorbar).

![image](https://user-images.githubusercontent.com/18143289/64261812-65ce2f00-cf2d-11e9-9950-1fa5dcd6ac2c.png)

larsoner commented:
Is it possible to set the background color of the colorbar to middle gray?

If not, is it possible to set the colorbar to a different set of colors than those on the brain? Namely those alpha-blended (by us in NumPy) with middle gray?

Also one thing I've always *really* wanted but seen no way to do in Mayavi is set the colorbar ticks. In particular, we probably want to default to setting them to the `fmin, fmid, fmax` values (pos and neg for two-sided). But even if there is a `pyvista` command for doing these things it would be great. Sometimes the colormap is a discrete thing (like subject count) and it would be good to be able to set that to `range(5, 20)` centered on the discrete things, for example. But not sure if this is possible with `pyvista`...

GuillaumeFavelier commented:
> Is it possible to set the background color of the colorbar to middle gray?

By disabling the *opacity* in the scalar bar, I obtain the following:

![image](https://user-images.githubusercontent.com/18143289/64264840-3cfc6880-cf32-11e9-9b97-c0fdad9c08bb.png)

Is it good enough?

> If not, is it possible to set the colorbar to a different set of colors than those on the brain? Namely those alpha-blended (by us in NumPy) with middle gray?

Technically, this is already the case. The colorbar uses scalar values plus a given colormap whereas the colors displayed on the brain come from a lookup table. An example by changing the colormap:

![image](https://user-images.githubusercontent.com/18143289/64265208-cca21700-cf32-11e9-9d54-527cf1eb9aed.png)

> Also one thing I've always really wanted but seen no way to do in Mayavi is set the colorbar ticks.

By looking at the [reference](https://docs.pyvista.org/plotting/plotting.html?highlight=scalar_bar#pyvista.BasePlotter.add_scalar_bar) I don't think it is available at the moement but I can eventually open an issue on PyVista.

larsoner commented:
> Is it good enough?

No, I don't think so. The fact that the `mne_analyze` colormap goes to gray in the middle is what makes it look correct here, but in general it won't be the case. For example consider `viridis` that is alpha=0 on the left and alpha=1 in the middle (i.e., what happens if you use `cmap='viridis', transparent=True`) will just look like `alpha=1.` everywhere, which is a misrepresentation of the colorbar people actually see.

What we want to do is take the colormap, alpha-blend with middle gray ourselves, then set that to be the colorbar.

But I also suspect that your `mne_analyze` colorbar is also incorrect. It's probably using a uniform spacing of the colorbar even for non-uniform `fmin/fmid/fmax`. Basically the PySurfer output is correct in that there is a large section in the middle that is pure gray (via alpha blending numpy arrays) and "squished" fmin->fmid sections on either end of the colorbar. Make sense? If not, we can talk it over on `gitter`. But if you think about how fmin/fmid/fmax can effectively compress and expand different parts of the colorbar (here in two-sided mode) it will hopefully make sense.

GuillaumeFavelier commented:
> What we want to do is take the colormap, alpha-blend with middle gray ourselves, then set that to be the colorbar.

This is very clear, thanks!

> But I also suspect that your mne_analyze colorbar is also incorrect. It's probably using a uniform spacing of the colorbar even for non-uniform fmin/fmid/fmax.

It's probably the case since I used the default parameters. I will work on it to get something better.

larsoner commented:
BTW there should already be code for doing all of these gymnastics in MNE and/or PySurfer

GuillaumeFavelier commented:
Indeed, I found it in the `scale_data_colormap()` function of PySurfer.

GuillaumeFavelier commented:
This is the experimental version, I hope it is better:

![image](https://user-images.githubusercontent.com/18143289/64360273-70122b00-d00a-11e9-872d-06dc9b5b1e52.png)

larsoner commented:
Yes much better, except that it appears to use too few colors (64?) instead of 256.

GuillaumeFavelier commented:
Latest result:

Mayavi/PySurfer | Mayavi backend | PyVista backend
---------------------|----------------------|----------------------
![image](https://user-images.githubusercontent.com/18143289/64443415-1e34d800-d0d2-11e9-8a60-57dd3f78fc9f.png) | ![image](https://user-images.githubusercontent.com/18143289/64443106-859e5800-d0d1-11e9-97d0-d273f7129b3e.png) | ![image](https://user-images.githubusercontent.com/18143289/64443036-58ea4080-d0d1-11e9-9ae2-6c6523a4b043.png)

**Note**

The Mayavi backend needs to some refactor: the colorbar is not valid and the `text_label` is out of focus. Also the dimensions of the window needs to be checked.

GuillaumeFavelier commented:
The colorbar and text label are now fixed for the Mayavi backend:

Mayavi/PySurfer | Mayavi backend | PyVista backend
---------------------|---------------------|-----------------------
![image](https://user-images.githubusercontent.com/18143289/64628215-47ff3f00-d3f1-11e9-85d0-a658b504f09d.png) | ![image](https://user-images.githubusercontent.com/18143289/64628265-5e0cff80-d3f1-11e9-9be3-860fd7ab00d6.png) | ![image](https://user-images.githubusercontent.com/18143289/64628319-75e48380-d3f1-11e9-9afa-5d48e592c61e.png)

Before using another example, I will focus on the CIs.

larsoner commented:
Awesome!

This gives me an idea -- we need a public function to make Mayavi-esque colorbars in matplotlib. You have presumably refactored code to do this. WDYT about pulling that code and making some public function `mne.viz.plot_colorbar(ax, colormap, clim, transparent=True, orientation='vertical')` or so that plots the 3D viz colorbar obtained for the given param combo, filling the given `ax`? You can hopefully see why it would be useful here:

https://github.com/mne-tools/mne-python/pull/6745

If this seems reasonable, feel free to push commits directly to that branch.

Otherwise, maybe we should wait until this branch is merged, then tackle the public function in #6745 since there is no rush on it. But I'm a bit worried that this PR is going to take a while to get in because it's going to have to get a lot of things right to be usable.

GuillaumeFavelier commented:
Recent development:

Mayavi/PySurfer | Mayavi backend | PyVista backend
---------------------|----------------------|---------------------
![image](https://user-images.githubusercontent.com/18143289/64969729-c4d46200-d8a4-11e9-86d2-43fea67d4c10.png) | ![image](https://user-images.githubusercontent.com/18143289/64969883-09f89400-d8a5-11e9-8e7c-c8a736fc06ee.png) | ![image](https://user-images.githubusercontent.com/18143289/64969811-eb929880-d8a4-11e9-8939-f9020cc53e2b.png)

I will now work to improve the camera settings.

larsoner commented:
Awesome!

agramfort commented:
beautiful !

>

GuillaumeFavelier commented:
Moving on to `tutorials/source-modeling/plot_object_source_estimate.py` (to demonstrate `add_foci()`):

Mayavi/PySurfer | PyVista backend
---------------------|---------------------
![image](https://user-images.githubusercontent.com/18143289/65138050-a1c3c280-da0a-11e9-9a16-d8847622486b.png) | ![image](https://user-images.githubusercontent.com/18143289/65138088-af794800-da0a-11e9-877b-0548d61dfe19.png)

**Note**

The text is missing since `add_text()` is not implemented yet and the colormap is slightly different.

agramfort commented:
it's also darker

GuillaumeFavelier commented:
You mean the texture on the brain?

agramfort commented:
yes. Lights seem lighter

>

GuillaumeFavelier commented:
Yes indeed. I noted your comment. I think we can definitely work on lighting to get something better perception-wise.

GuillaumeFavelier commented:
But for the colormap, I still think I missed a parameter or post-processing somewhere. I just need to find it

GuillaumeFavelier commented:
With fixed colormap:

Mayavi/PySurfer | Mayavi backend | PyVista backend
---------------------|----------------------|---------------------
![image](https://user-images.githubusercontent.com/18143289/65426073-d327fe00-de0f-11e9-87a3-6187e4c68ead.png) | ![image](https://user-images.githubusercontent.com/18143289/65426257-226e2e80-de10-11e9-8a5c-91c1ebdf31ac.png) | ![image](https://user-images.githubusercontent.com/18143289/65426340-534e6380-de10-11e9-9d5f-3f8afe180841.png)

GuillaumeFavelier commented:
With subplots:

Mayavi backend | PyVista backend
---------------------|----------------------
![image](https://user-images.githubusercontent.com/18143289/65436548-9a455480-de22-11e9-87ef-f2160f8055dc.png) | ![image](https://user-images.githubusercontent.com/18143289/65436694-d24c9780-de22-11e9-9149-6e1702a63761.png)

GuillaumeFavelier commented:
Now main blockers would be camera settings, `TimeViewer`, some API changes or bugs. And I need to trigger `[circle full]` to check the results.

agramfort commented:
I am crying.... :)

>

GuillaumeFavelier commented:
The latest work on `show_view()` and the parameters of the camera could drastically improve the default point of view of the scene. Support in case of `hemi=split` is not supported yet but it is in the todo list.

GuillaumeFavelier commented:
The screenshots with the default camera settings:

![image](https://user-images.githubusercontent.com/18143289/65525033-26bd4900-deef-11e9-94fd-14520bccdb24.png)
![image](https://user-images.githubusercontent.com/18143289/65525093-3fc5fa00-deef-11e9-93c7-8d5786794a59.png)
![image](https://user-images.githubusercontent.com/18143289/65525168-5cfac880-deef-11e9-9a2e-2732f64ba1bf.png)
![image](https://user-images.githubusercontent.com/18143289/65525213-73a11f80-deef-11e9-881a-0166b3b7627e.png)

larsoner commented:
@GuillaumeFavelier feel free to rebase now that #6819 is in

banesullivan commented:
Dang, this is cool to see how PyVista is being used!! 

FYI, https://github.com/pyvista/pyvista/pull/380 which made it into the latest PyVista release might be useful to you all for setting the layout of a multi-scene PyVista window

GuillaumeFavelier commented:
I'm not sure but I think our users prefer to have each hemisphere on opposite sides. However, maybe using [linked views](https://docs.pyvista.org/examples/02-plot/linked.html) could be a nice addition maybe? what do you think @agramfort @larsoner ?

agramfort commented:
you want to see both sides for one brain / one activation. As you want to
see all without having to rotate when you scroll in time.

you will want linked views if you want to see different conditions or
different subjects with the same angle and at the same time.
btw this would be really awesome to have.

>

GuillaumeFavelier commented:
Next step is to fix `AttributeError: 'NoneType' object has no attribute 'keys'` happening for a few tutorials.

GuillaumeFavelier commented:
This is the current result: https://15710-1301584-gh.circle-artifacts.com/0/dev/index.html

larsoner commented:
Most look reasonable!

- [This one](https://15710-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_mne_dspm_source_localization.html) suggests `plot_vector_source_estimates` still needs to be changed
- Colorbar tick marks are too small/hard to read (looks like you've noted this already)
- Should probably use a larger, sans-serif, non-monospaced font for colorbar (but unlike Mayavi, leave non-italic) -- feel free to tweak the description above
- Cylinder size discrepancy (radius vs diameter, maybe?): https://mne.tools/dev/auto_tutorials/source-modeling/plot_eeg_no_mri.html vs https://15710-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_eeg_no_mri.html
- Double-colorbar problem: https://15710-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/stats-source-space/plot_stats_cluster_spatio_temporal.html (looks like you've noted this already)

agramfort commented:
Awesome 👏 

Just a few glitches remain but you’re almost there !

GuillaumeFavelier commented:
The latest results are available: https://15924-1301584-gh.circle-artifacts.com/0/dev/index.html

@larsoner I can take of `plot_vector_estimates` in a different PR, this one is more focused on `plot_source_estimates`. I think I addressed most of your remarks.

Colorbar ticks are bigger and I changed the default font:

![image](https://user-images.githubusercontent.com/18143289/66123686-ee9ac200-e5e2-11e9-8bb4-0658e4d338c6.png)

The 'double colorbars' bug occuring when using both hemispheres has been fixed too:

![image](https://user-images.githubusercontent.com/18143289/66123860-3f121f80-e5e3-11e9-8e08-07fb73c301c8.png)

We pull PyVista from master for now to get the latest fixes.

And for cylinder vs radius, I didn't manage to find out how Mayavi decide on its automatic radius size for now.

larsoner commented:
@GuillaumeFavelier looks great!

So is this ready for review/merge from your end?

GuillaumeFavelier commented:
Yes it is! Please let me know if you have any feedbacks or remarks.

larsoner commented:
Can you update the tickboxes in the top comment? Either check them off or say what the plan is (e.g., deal with them later or never)

GuillaumeFavelier commented:
Okay it's updated now. So everything in the **ToDo in a follow-up PR** and **Feature request** sections can be done afterwards I think.

Among the bugs: 
- 'Fix screenshot with mode='rgba' for the pyvista backend': it's a note to myself, we can delay it as long as it's not needed.
- 'Difference in cylinder radius for quiver3d': I need help or more time to find the automatic size chosen by Mayavi.

For the rest, I think that we can adjust the prioritydepending on the usage basically.

GuillaumeFavelier commented:
At least I hope I didn't miss anything critical among the features provided by `Brain`

larsoner commented:
Maybe one of these provides the default:

https://github.com/enthought/mayavi/blob/3c90ebb414ccd70b6bd25b69ce192b9177dd5903/mayavi/components/glyph_source.py#L256-L257

https://github.com/enthought/mayavi/blob/5c2694b72b329b8d5c469bc459a211378e2c8581/tvtk/pyface/actors.py#L68-L72

https://github.com/enthought/mayavi/blob/5c2694b72b329b8d5c469bc459a211378e2c8581/tvtk/tools/visual.py#L1213

Found by searching "cylinder" then "cylindersource" in the Mayavi repo. I would try writing a little script to make a cylinder actor, then change each of these one by one to see which one (hopefully exactly one) of them actually changes the default params.

But also make sure that `scale_factor` is set otherwise it does some magic:

https://github.com/enthought/mayavi/blob/b3fc35218dda9776d8f1a407663bfb49783dca12/mayavi/tools/helper_functions.py#L174-L180

larsoner commented:
Running:
```
import matplotlib.pyplot as plt
from mayavi import mlab
mlab.quiver3d(0, 0, 0, 0, 0, 1, scale_mode='vector', mode='cylinder')
```
changing line 256 of `mayavi/components/glyph_source.py` from:
```
             'cylinder_source': tvtk.CylinderSource(height=1.0, radius=0.15,
```
which produces

<img width="512" alt="Screen Shot 2019-10-03 at 08 18 30" src="https://user-images.githubusercontent.com/2365790/66126164-d1e89500-e5b6-11e9-8f34-9b46cdc93f31.png">

to
```
             'cylinder_source': tvtk.CylinderSource(height=1.0, radius=1.15,
```
I get

<img width="512" alt="Screen Shot 2019-10-03 at 08 19 41" src="https://user-images.githubusercontent.com/2365790/66126171-d57c1c00-e5b6-11e9-8d35-8385719becb7.png">

So I think the default is just 0.15 (presumably times the vector length), at least given `scale_mode='vector'`.

agramfort commented on mne/viz/_brain/utils.py:
```diff
@@ -0,0 +1,94 @@
+
+    Parameters
+    ----------
+    faces : array of shape [n_triangles x 3]
```
```suggestion
    faces : array, shape (n_triangles, 3)
```

agramfort commented on mne/viz/_brain/utils.py:
```diff
@@ -0,0 +1,94 @@
+
+       If smoothing_steps is None, as many smoothing steps are applied until
+       the whole mesh is filled with with non-zeros. Only use this option if
+       the vertices correspond to a subsampled version of the mesh.
```
indent pb

agramfort commented on mne/viz/_brain/utils.py:
```diff
@@ -0,0 +1,94 @@
+    smoothing_steps : int or None
+        number of smoothing steps (Default: 20)
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see surfer.verbose).
```
%(verbose)s

GuillaumeFavelier commented:
I couldn't reproduce, this is what I get:

![image](https://user-images.githubusercontent.com/18143289/66133450-f6646180-e5f6-11e9-8ca3-279163521bef.png)

Rendered without the toolbar:

![image](https://user-images.githubusercontent.com/18143289/66133565-2dd30e00-e5f7-11e9-9346-bbd345ed274d.png)

larsoner commented:
@GuillaumeFavelier I do not have Alex's problem but I do have what appears to be a HiDPI problem -- the location of the text that should be all the way to the right is only halfway over:

<img width="912" alt="Screen Shot 2019-10-03 at 11 38 37" src="https://user-images.githubusercontent.com/2365790/66141893-804e0380-e5d2-11e9-8725-c1ee533e93d0.png">

And all the font scalings are off by a factor of 2. So somewhere in your code or `pyvista` they need to scale by the physical-to-logical pixel ratio in things like the text size and the text positioning in terms of "pixels".

GuillaumeFavelier commented:
The latest changes should fix the issues with the tube radius and the text position. I started Circle again for one last build hopefully.

larsoner commented:
Agreed, awesome work @GuillaumeFavelier !

larsoner commented:
Just noticed that screenshots are captured correctly for multi-view scenes now:

https://16063-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/simulation/plot_point_spread.html

This is great!

agramfort commented:
Congrats @GuillaumeFavelier ! This is a huge milestone

GuillaumeFavelier commented:
Some news about the scalar bar ticks with VTK @larsoner (following your https://github.com/mne-tools/mne-python/pull/6731#issuecomment-527920541):

@banesullivan reported the issue on [vtk.discourse.org](https://discourse.vtk.org/t/tick-lines-with-vtkscalarbaractor/2268) and you can follow the progress [on this ticket](https://gitlab.kitware.com/vtk/vtk/-/issues/17751).

(Sorry for the delay on this).

larsoner commented:
That looks like it has to do with whether ticks and labels are drawn, rather than being able to specify the tick+label numerical values/locations
